Filters for text searching and statuses

joj0s opened the pull request:
This PR adds a text filter on top of the "Browse Traits" page that filters by trait name, ontology labels and ontology CURIE. It also adds buttons for selecting each distinct status of a trait. 

Compared to the UI mockup, I excluded the "FIlter" button next to the textbox since it is all done instantly using JavaScript and we don't need a button to trigger a request.

Closes #16, closes #17, closes #20

joj0s commented:
@tskir  I have also added colored labels to the trait table and links to to point the CURIE column to its appropriate URI. if you like the changes, we can also close #20

tskir commented on traitcuration/static/js/table-filtering.js:
```diff
@@ -0,0 +1,33 @@
+  table.search(searchstring, ["name", "curie", "label"]);
+}
+
+function filterClicked(status, activeClass) {
```
Could you add some documentation to this script? I'm thinking one top-level description (like a docstring) + some comments throughout the code, at least for the major branching points. They don't have to be verbose

tskir commented on traitcuration/traits/views.py:
```diff
@@ -0,0 +1,33 @@
+    return render(request, 'traits/trait_detail.html', context)
+
+
+def get_status_dict(traits):
```
So concerning the `views.py` module, do you suppose it would be better to leave only the actual views in it (e.g. something which returns `render(...)`, and move helper routines, such as `get_status_dict`, into a separate module?

joj0s replied:
Yeah, I also thought about it and it would be cleaner. I moved it into a `utils.py` module

tskir commented on traitcuration/static/js/listjs/list.js:
```diff
@@ -0,0 +1,2 @@
+// List.js v1.4.1 (http://www.listjs.com) by Jonny Strömberg (http://javve.com)
```
The same comment applies to this library, depending on which decision we will go with in #35: either use external CDN, or install via Yarn, or keep locally and include update/installation instructions

joj0s commented:
> The only small bug is that the behaviour of the Enter button is somewhat confusing: when a user presses it while focused on the search field, anything the user has entered there will get erased and they will be redirected to the front page. Enter should not do this.

You are right, I didn't test the Enter button functionality. I will fix that
 
>Another consequence of instant search is that it's not possible to bookmark a combination of specific filters, as there isn't a dedicated URL to use for that. If this can be fixed easily, then please do so as part of this PR. Otherwise, we can create a separate low-priority issue in the backlog for that.

Yes, that is the drawback of using JavaScript to filter. We cannot include the filters in a URL so I think we should add a new issue for that, although I am not quite sure what that would be. Possibly a page with query history?

joj0s commented:
Regarding the second iteration of changes in this PR:

- Created a `utils.py` file and moved the helper function from views.py to there
- Documented the JS script for filtering
- Fixed the Enter button functionality
- Added loading list.js library via CDN
